src/python/test/test_locking.py

```diff
@@ -0,0 +1,167 @@
+            self.start_event.set()
+            self.stop_event.wait()
+            fcntl.lockf(fd, fcntl.LOCK_UN)
+
+    def __enter__(self):
+        self.start()
+        self.start_event.wait()
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.stop_event.set()
+        self.join()
+
+
+class WriterProcess(mp.Process):
+    """
+    Try to trigger a race condition by synchronising multiple processes writing
+    to the same dataset
+    """
+    def __init__(self, filepath, ds_name, offset, data, ready_event, start_event):
+        super(WriterProcess, self).__init__()
+        self.filepath = filepath
+        self.ds_name = ds_name
+        self.offset = offset
+        self.data = data
+        self.ready_event = ready_event
+        self.start_event = start_event
+        self.daemon = True
+
+    def run(self):
+        dataset = z5py.File(self.filepath)[self.ds_name]
+        slicing = tuple(slice(o, o + s) for o, s in zip(self.offset, self.data.shape))
+        self.ready_event.set()
+        self.start_event.wait()
+        dataset[slicing] = self.data
+
+
+class TestLockfManager(unittest.TestCase):
+    """Check that the lockf utility defined here actually works"""
+    def setUp(self):
+        self.logger = logging.getLogger(self.id())
+        self.fpath = "test.file"
+        with open(self.fpath, 'w') as f:
+            f.write("some text")
+
+    @unittest.skipUnless(fcntl, "Locking is only supported on unix")
+    def test_locks_respected(self):
+        logger.debug("Getting first lock")
+        with OtherLock(self.fpath):
+            with self.assertRaises(IOError):
+                logger.debug("Attempting to open file for writing")
+                with open(self.fpath) as f:
+                    f.write("some text which should not be written")
+
+
+@add_metaclass(ABCMeta)
+class LockingTestMixin(object):
+    def setUp(self):
+        self.shape = (100, 100, 100)
+        self.fpath = 'array.' + self.data_format
+        self.file = z5py.File(self.fpath, use_zarr_format=self.data_format == 'zarr')
+        self.dtype = np.uint64
+        self.ds_name = "ds"
+        self.dataset = self.file.create_dataset(
+            self.ds_name, data=np.ones(self.shape, dtype=self.dtype), chunks=self.shape
+        )
+        self.block_path = z5py.util.get_block_path(self.dataset, [0, 0, 0])
+        np.random.seed(SEED)
+        self.logger = logging.getLogger(self.id())
+
+    def tearDown(self):
+        try:
+            rmtree('array.' + self.data_format)
+        except OSError:
+            pass
+
+    @unittest.skipUnless(fcntl, "Locking only supported on unix")
+    def test_lockf_respected(self):
+        with self.assertRaises(Exception) as captured:
+            with OtherLock(self.block_path):
+                self.logger.debug("Attempting to write to block")
+                self.dataset[:] = np.ones(self.dataset.shape, dtype=self.dtype) * 2
+        print(captured.exception)
```

Comment:
Not sure what kind of exception the C++ layer will throw; this should be made more specific.